Refactor tests that call pip, to use local test specific wheel file

jezdez opened the pull request:
- Updated test_create.py and test_main_export.py to replace 'small_python_package' with 'small-python-package' for consistency in package naming.
- Removed unnecessary imports and adjusted assertions to enhance clarity and maintainability.
- Improved pip package handling in tests by using 'flask' as a reliable dependency, ensuring consistent behavior across test scenarios.
- Enhanced test coverage for environment export functionality, particularly regarding pip dependencies and channel handling.

Follow up to #14886.

<!-- Hello! Thanks for submitting a PR! To help make things go a bit more
     smoothly, we would appreciate it if you follow this template. -->

### Description

<!-- Good things to put here include:
       - reasons for the change (please link any relevant issues!),
       - any noteworthy (or hacky) choices to be aware of,
       - or what the problem resolved here looked like. -->

### Checklist - did you ...

<!-- If any of the following items aren't relevant to your contribution,
     please either tick them or use ~strikethrough~ so we know you've gone
     through the checklist. -->

- ~[ ] Add a file to the `news` directory ([using the template](https://github.com/conda/conda/blob/main/news/TEMPLATE)) for the next release's release notes?~
     <!-- All "significant" changes should get an entry:
            - user-facing changes or enhancements
            - bug fixes
            - deprecations
            - documentation updates
            - etc -->
- [x] Add / update necessary tests?
- ~[ ] Add / update outdated documentation?~


<!-- Just as a reminder, everyone in all conda org spaces (including PRs)
     must follow the Conda Org Code of Conduct (link below).

     Finally, once again, thanks for your time and effort. If you have any
     feedback in regards to your experience contributing here, please
     let us know!

     Helpful links:
       - Conda Org COC: https://github.com/conda/conda/blob/main/CODE_OF_CONDUCT.md
       - Contributing docs: https://github.com/conda/conda/blob/main/CONTRIBUTING.md -->

jezdez committed:
Refactor tests to improve package handling and consistency

- Updated test_create.py and test_main_export.py to replace 'small_python_package' with 'small-python-package' for consistency in package naming.
- Removed unnecessary imports and adjusted assertions to enhance clarity and maintainability.
- Improved pip package handling in tests by using 'flask' as a reliable dependency, ensuring consistent behavior across test scenarios.
- Enhanced test coverage for environment export functionality, particularly regarding pip dependencies and channel handling.

These changes streamline the testing process and ensure accurate representation of package specifications in tests.

dholth commented on conda/testing/fixtures.py:
```diff
@@ -306,6 +308,110 @@ def session_conda_cli() -> Iterator[CondaCLIFixture]:
+        *argv: str | os.PathLike[str] | Path,
+        prefix: str | os.PathLike[str] | Path,
+        raises: type[Exception] | tuple[type[Exception], ...],
+    ) -> tuple[str, str, ExceptionInfo]: ...
```
```
>>> isinstance(pathlib.Path(), os.PathLike)
True
```

jezdez committed:
Refactor prefix handling to use PathType for improved type consistency

- Updated the `prefix` parameter in `Index` and `ReducedIndex` classes to use `PathType` instead of a combination of types including `os.PathLike` and `Path`.
- Modified the `prefix_path` parameter in `PrefixData` class to utilize `PathType` for better type clarity.
- Adjusted the `argv` parameter in `CondaCLIFixture` and `PipCLIFixture` classes to accept `PathType`, ensuring consistent handling of path-like arguments.
- Enhanced the `_cast_args` method in `CondaCLIFixture` to reflect the new type for argument casting.

These changes streamline type handling across the codebase, improving maintainability and reducing potential type-related issues.

jezdez committed:
Merge branch 'main' into wheel-exporter-hook

jezdez committed:
Refactor test_get_packages_behavior_with_interoperability to improve clarity

- Updated the test to remove the unused 'test_recipes_channel' parameter, simplifying the function signature.
- Changed the wheel path to use 'wheelhouse' instead of 'test_recipes_channel' for better context and consistency in package handling.

These changes enhance the maintainability and readability of the test code.

jezdez committed:
Merge branch 'wheel-exporter-hook' of github.com:conda/conda into wheel-exporter-hook

jezdez committed:
Fix typo in test_get_packages_behavior_with_interoperability to use hyphenated package name

- Updated the assertion for the Python package name from 'small_python_package' to 'small-python-package' for consistency with naming conventions.

This change improves clarity in the test and aligns with previous updates to package naming.

kenodegard commented on conda/testing/fixtures.py:
```diff
@@ -306,6 +310,110 @@ def session_conda_cli() -> Iterator[CondaCLIFixture]:
+                # python executable not found
+                raise RuntimeError(
+                    f"Python not found in environment {prefix_path}: {python_exe}"
+                ) from e
```
```suggestion
                )
```

kenodegard commented on conda/testing/fixtures.py:
```diff
@@ -306,6 +310,110 @@ def session_conda_cli() -> Iterator[CondaCLIFixture]:
+class PipCLIFixture:
+    """Fixture for calling pip in specific conda environments."""
+
+    capsys: CaptureFixture | None
```
do we actually need `capsys` since this is a `subprocess.run(capture_output=True)`?

kenodegard commented on tests/cli/test_main_export.py:
```diff
@@ -653,19 +651,27 @@ def test_export_pip_dependencies_handling(conda_cli, format_name, parser_func):
+
+        # Install using pip_cli fixture for better error handling
+        pip_stdout, pip_stderr, pip_code = pip_cli(
+            "install", str(wheel_path), prefix=prefix
```
```suggestion
            "install", wheel_path, prefix=prefix
```

kenodegard commented on tests/cli/test_main_export.py:
```diff
@@ -785,15 +791,18 @@ def test_export_override_channels_and_ignore_channels_independence(conda_cli):
-            shell=True,
+        wheel_path = wheelhouse / "small_python_package-1.0.0-py3-none-any.whl"
+        pip_stdout, pip_stderr, pip_code = pip_cli(
+            "install", str(wheel_path), prefix=prefix
```
```suggestion
            "install", wheel_path, prefix=prefix
```

kenodegard commented on tests/test_create.py:
```diff
@@ -823,25 +824,29 @@ def test_strict_resolve_get_reduced_index(monkeypatch: MonkeyPatch):
-            shell=True,
+        wheel_path = wheelhouse / "small_python_package-1.0.0-py3-none-any.whl"
+        pip_stdout, pip_stderr, pip_code = pip_cli(
+            "install", str(wheel_path), prefix=prefix
```
```suggestion
            "install", wheel_path, prefix=prefix
```

kenodegard commented on tests/test_create.py:
```diff
@@ -853,26 +858,30 @@ def test_list_with_pip_no_binary(tmp_env: TmpEnvFixture, conda_cli: CondaCLIFixt
-            shell=True,
+        wheel_path = wheelhouse / "small_python_package-1.0.0-py3-none-any.whl"
+        pip_stdout, pip_stderr, pip_code = pip_cli(
+            "install", str(wheel_path), prefix=prefix
```
```suggestion
            "install", wheel_path, prefix=prefix
```

kenodegard commented on tests/test_create.py:
```diff
@@ -1734,26 +1742,17 @@ def test_conda_pip_interop_pip_clobbers_conda(
-            "list",
-            "--format=freeze",
-        )
+        stdout, _, _ = pip_cli("list", "--format=freeze", prefix=prefix)
```
much nicer!

jezdez committed:
Apply suggestions from code review

Co-authored-by: Ken Odegard <kodegard@anaconda.com>

jezdez committed:
Refactor PipCLIFixture to remove capsys dependency

- Removed the capsys parameter from PipCLIFixture and related functions to simplify the interface.
- Updated pip_cli fixture to instantiate PipCLIFixture without capsys.
- Adjusted test cases to reflect the changes in the pip_cli function signature.

These modifications enhance the clarity and maintainability of the testing fixtures.

jezdez committed:
Update tests/cli/test_main_export.py

Co-authored-by: Ken Odegard <kodegard@anaconda.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jezdez committed:
Update conda/testing/fixtures.py

Co-authored-by: Ken Odegard <kodegard@anaconda.com>
